aws/resource_aws_wafv2_web_acl.go

```diff
@@ -0,0 +1,843 @@
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.DeleteWebACL(r)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error deleting WAFv2 WebACL: %s", err)
+	}
+
+	return nil
+}
+
+func wafv2WebACLRootStatementSchema(level int) *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Required: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"and_statement":                         wafv2StatementSchema(level - 1),
+				"byte_match_statement":                  wafv2ByteMatchStatementSchema(),
+				"geo_match_statement":                   wafv2GeoMatchStatementSchema(),
+				"ip_set_reference_statement":            wafv2IpSetReferenceStatementSchema(),
+				"managed_rule_group_statement":          wafv2ManagedRuleGroupStatementSchema(),
+				"not_statement":                         wafv2StatementSchema(level - 1),
+				"or_statement":                          wafv2StatementSchema(level - 1),
+				"rate_based_statement":                  wafv2RateBasedStatementSchema(level - 1),
+				"regex_pattern_set_reference_statement": wafv2RegexPatternSetReferenceStatementSchema(),
+				"rule_group_reference_statement":        wafv2RuleGroupReferenceStatementSchema(),
+				"size_constraint_statement":             wafv2SizeConstraintSchema(),
+				"sqli_match_statement":                  wafv2SqliMatchStatementSchema(),
+				"xss_match_statement":                   wafv2XssMatchStatementSchema(),
+			},
+		},
+	}
+}
+
+func wafv2ManagedRuleGroupStatementSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"excluded_rule": wafv2ExcludedRuleSchema(),
+				"name": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringLenBetween(1, 128),
+				},
+				"vendor_name": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringLenBetween(1, 128),
+				},
+			},
+		},
+	}
+}
+
+func wafv2ExcludedRuleSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"name": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringLenBetween(1, 128),
+				},
+			},
+		},
+	}
+}
+
+func wafv2RateBasedStatementSchema(level int) *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				// Required field but currently only supports "IP"
+				"aggregate_key_type": {
+					Type:     schema.TypeString,
+					Optional: true,
+					Default:  wafv2.RateBasedStatementAggregateKeyTypeIp,
```

Comment:
if required per AWS docs, we could alternatively remove the default but validate this field with a `ValidateFunc: validation.StringInSlice()` in case AWS later adds different aggregate types